File: public/desk/js/input.js
// Your mouse and keyboard → the other computer.
//
// Pointer: taking control locks the pointer (Pointer Lock API), which gives
// unbounded relative motion and hides the local cursor. We integrate that into
// our own position and send it as an ABSOLUTE normalized point, so a dropped
// frame can never leave the two cursors permanently offset.
//
// Keyboard: the browser hands us `event.key` with the layout, dead keys and
// AltGr composition ALREADY applied — so printable characters travel as text
// (the agent injects them layout-correctly) and only non-character keys and
// chords travel as key events. Modifiers are pressed on the host lazily, right
// before a chord or a click needs them: holding Shift/AltGr over there while
// text is being injected would corrupt what you typed.
//
// Right Ctrl is the local menu key (never forwarded): tap = take/give back
// control, and its chords drive this page — including the chords no browser
// would ever let us capture (Ctrl+Alt+Del, Alt+Tab, Super).
const ZRDeskInput = (() => {
  let stage = null, send = () => {}, view = null, onAction = () => {};
  let driving = false, keyHold = false, raw = false;
  let nx = .5, ny = .5, pendingMove = false, rafId = 0;
  let scrollAcc = 0, scrollAccX = 0;
  let textBuf = '';
  const heldKeys = new Set();
  const heldButtons = new Set();
  const sentMods = new Set();
  let menuHeld = false, menuUsed = false;
  let keyboardLocked = false;

  const BUTTONS = { 0: 'left', 1: 'middle', 2: 'right' };

  // browser key → the names the agent's injector knows (robotgo keyNames)
  const KEYS = {
    Enter: 'enter', NumpadEnter: 'enter', Escape: 'escape', Tab: 'tab',
    Backspace: 'backspace', Delete: 'delete', Insert: 'insert',
    Home: 'home', End: 'end', PageUp: 'pageup', PageDown: 'pagedown',
    ArrowUp: 'up', ArrowDown: 'down', ArrowLeft: 'left', ArrowRight: 'right',
    CapsLock: 'capslock', PrintScreen: 'printscreen', ContextMenu: 'menu',
    NumLock: 'num_lock',
  };
  const CODES = {
    Numpad0: 'num0', Numpad1: 'num1', Numpad2: 'num2', Numpad3: 'num3',
    Numpad4: 'num4', Numpad5: 'num5', Numpad6: 'num6', Numpad7: 'num7',
    Numpad8: 'num8', Numpad9: 'num9', NumpadAdd: 'num_plus',
    NumpadSubtract: 'num_minus', NumpadMultiply: 'num_asterisk',
    NumpadDivide: 'num_slash',
  };

  function keyNameOf(e) {
    if (KEYS[e.key]) return KEYS[e.key];
    if (CODES[e.code]) return CODES[e.code];
    if (/^F([1-9]|1[0-2])$/.test(e.key)) return e.key.toLowerCase();
    return null;
  }

  function config(opts) {
    stage = opts.stage;
    send = opts.send || send;
    view = opts.view;
    onAction = opts.onAction || onAction;

    stage.addEventListener('mousedown', onMouseDown);
    window.addEventListener('mouseup', onMouseUp);
    window.addEventListener('mousemove', onMouseMove);
    stage.addEventListener('wheel', onWheel, { passive: false });
    stage.addEventListener('contextmenu', (e) => e.preventDefault());
    window.addEventListener('keydown', onKeyDown, true);
    window.addEventListener('keyup', onKeyUp, true);
    window.addEventListener('blur', () => { if (driving) release(); });
    document.addEventListener('pointerlockchange', onLockChange);
    document.addEventListener('pointerlockerror', () => onAction('lockfail'));
    document.addEventListener('fullscreenchange', onFullscreenChange);
  }

  // ── taking / giving back control ──────────────────────────────────────────
  function take() {
    if (!stage) return;
    // unadjustedMovement asks for raw device deltas (no OS pointer acceleration)
    // so a flick here matches a flick over there; older browsers ignore it.
    const req = stage.requestPointerLock({ unadjustedMovement: true });
    if (req && typeof req.catch === 'function') req.catch(() => stage.requestPointerLock());
  }
  function release() { if (document.pointerLockElement) document.exitPointerLock(); else setDriving(false); }
  function toggle() { driving ? release() : take(); }

  function onLockChange() { setDriving(document.pointerLockElement === stage); }

  function setDriving(on) {
    if (driving === on) return;
    driving = on;
    if (!on) releaseAll();
    view && view.setCursor(nx, ny, on);
    onAction(on ? 'grab' : 'ungrab');
  }

  // Never leave a key or button held on the other machine.
  function releaseAll() {
    heldButtons.forEach((b) => send({ t: 'up', b }));
    heldButtons.clear();
    heldKeys.forEach((k) => { if (keyHold) send({ t: 'kup', k }); });
    heldKeys.clear();
    releaseMods();
  }

  // ── pointer ───────────────────────────────────────────────────────────────
  function onMouseMove(e) {
    if (!driving) return;
    const r = view.rect();
    if (!r.w || !r.h) return;
    nx = clamp01(nx + e.movementX / r.w);
    ny = clamp01(ny + e.movementY / r.h);
    view.setCursor(nx, ny, true);
    queueMove();
  }

  // one move per animation frame: the host injects at most one per captured
  // frame anyway, and a flick shouldn't flood a mobile uplink
  function queueMove() {
    pendingMove = true;
    if (rafId) return;
    rafId = requestAnimationFrame(() => {
      rafId = 0;
      if (!pendingMove) return;
      pendingMove = false;
      send({ t: 'mvabs', nx: round6(nx), ny: round6(ny) });
    });
  }

  function onMouseDown(e) {
    if (!driving) return;   // the first click is what takes control (app.js)
    e.preventDefault();
    const b = BUTTONS[e.button];
    if (!b) return;
    syncMods(e);
    send({ t: 'mvabs', nx: round6(nx), ny: round6(ny) });   // pin before pressing
    send({ t: 'down', b });
    heldButtons.add(b);
  }

  function onMouseUp(e) {
    const b = BUTTONS[e.button];
    if (!b || !heldButtons.has(b)) return;
    e.preventDefault();
    send({ t: 'up', b });
    heldButtons.delete(b);
  }

  function onWheel(e) {
    if (!driving) return;
    e.preventDefault();
    // deltaMode: 0 = pixels, 1 = lines, 2 = pages → notches
    const div = e.deltaMode === 0 ? 100 : e.deltaMode === 1 ? 3 : 1;
    // browser: down is positive. robotgo: up is positive.
    scrollAcc += -(e.deltaY / div) * 3;
    scrollAccX += (e.deltaX / div) * 3;
    const dy = Math.trunc(scrollAcc), dx = Math.trunc(scrollAccX);
    if (!dx && !dy) return;
    scrollAcc -= dy; scrollAccX -= dx;
    send({ t: 'scroll', dx, dy });
  }

  // ── keyboard ──────────────────────────────────────────────────────────────
  function onKeyDown(e) {
    // Right Ctrl: the local menu key. Held = nothing is forwarded.
    if (e.code === 'ControlRight') {
      e.preventDefault();
      if (!menuHeld) { menuHeld = true; menuUsed = false; releaseAll(); }
      return;
    }
    if (menuHeld) { e.preventDefault(); chord(e); return; }
    if (!driving) return;
    e.preventDefault();

    if (e.repeat && !raw && isPrintable(e)) { queueText(e.key); return; }
    if (e.key === 'Dead' || e.key === 'Unidentified') return;   // composing
    if (isModifier(e.key)) return;                              // pressed lazily

    if (!raw && isPrintable(e)) {
      // the character the browser composed IS what the user typed — send it as
      // text so the host injects it in ITS layout
      releaseMods();
      queueText(e.key);
      return;
    }

    const name = raw ? rawName(e) : keyNameOf(e);
    if (!name) return;
    syncMods(e);
    if (keyHold) { send({ t: 'kdown', k: name }); heldKeys.add(name); }
    else send({ t: 'key', k: name, mods: modList(e) });   // agent < 1.7.0
  }

  function onKeyUp(e) {
    if (e.code === 'ControlRight') {
      e.preventDefault();
      menuHeld = false;
      if (!menuUsed) toggle();     // a bare tap flips control
      return;
    }
    if (!driving) return;
    e.preventDefault();
    if (isModifier(e.key)) { releaseModIfUp(e); return; }
    const name = raw ? rawName(e) : keyNameOf(e);
    if (!name || !heldKeys.has(name)) return;
    heldKeys.delete(name);
    if (keyHold) send({ t: 'kup', k: name });
  }

  // raw mode: forward the physical key, letting the other computer's layout
  // decide the character. Right for games, wrong for typing across layouts.
  function rawName(e) {
    const c = e.code;
    if (/^Key[A-Z]$/.test(c)) return c.slice(3).toLowerCase();
    if (/^Digit[0-9]$/.test(c)) return c.slice(5);
    if (CODES[c]) return CODES[c];
    const punct = {
      Minus: '-', Equal: '=', BracketLeft: '[', BracketRight: ']', Backslash: '\\',
      Semicolon: ';', Quote: "'", Comma: ',', Period: '.', Slash: '/', Backquote: '`',
      Space: 'space', IntlBackslash: '\\',
    };
    if (punct[c]) return punct[c];
    if (KEYS[e.key]) return KEYS[e.key];
    if (/^F([1-9]|1[0-2])$/.test(e.key)) return e.key.toLowerCase();
    const mods = { ShiftLeft: 'shift', ShiftRight: 'shift', ControlLeft: 'ctrl', AltLeft: 'alt', AltRight: 'alt', MetaLeft: 'cmd', MetaRight: 'cmd' };
    return mods[c] || null;
  }

  function isPrintable(e) {
    if (!e.key || e.key.length !== 1) return false;
    if (e.metaKey) return false;
    // AltGr reports as Ctrl+Alt on Windows, yet it composes a real character
    if (e.getModifierState && e.getModifierState('AltGraph')) return true;
    return !e.ctrlKey && !e.altKey;
  }
  function isModifier(k) { return k === 'Shift' || k === 'Control' || k === 'Alt' || k === 'Meta' || k === 'AltGraph'; }

  // text is batched per frame: typing fast shouldn't mean one frame per letter
  function queueText(ch) {
    textBuf += ch;
    if (textBuf.length === 1) requestAnimationFrame(flushText);
  }
  function flushText() {
    if (!textBuf) return;
    send({ t: 'text', s: textBuf });
    textBuf = '';
  }

  function modList(e) {
    const m = [];
    if (e.shiftKey) m.push('shift');
    if (e.ctrlKey) m.push('ctrl');
    if (e.altKey) m.push('alt');
    if (e.metaKey) m.push('meta');
    return m;
  }

  // press on the host exactly the modifiers held here, right before an event
  // that needs them
  function syncMods(e) {
    if (!keyHold) return;
    const want = new Set();
    if (e.shiftKey) want.add('shift');
    if (e.ctrlKey) want.add('ctrl');
    if (e.altKey) want.add('alt');
    if (e.metaKey) want.add('cmd');
    want.forEach((m) => { if (!sentMods.has(m)) { send({ t: 'kdown', k: m }); sentMods.add(m); } });
    sentMods.forEach((m) => { if (!want.has(m)) { send({ t: 'kup', k: m }); sentMods.delete(m); } });
  }

  function releaseModIfUp(e) {
    const name = { Shift: 'shift', Control: 'ctrl', Alt: 'alt', Meta: 'cmd' }[e.key];
    if (!name || !sentMods.has(name)) return;
    send({ t: 'kup', k: name });
    sentMods.delete(name);
  }

  function releaseMods() {
    sentMods.forEach((m) => { if (keyHold) send({ t: 'kup', k: m }); });
    sentMods.clear();
  }

  // ── Right Ctrl chords ─────────────────────────────────────────────────────
  function chord(e) {
    menuUsed = true;
    switch (e.code) {
      case 'KeyF': onAction('fullscreen'); break;
      case 'KeyP': onAction('quality'); break;
      case 'KeyM': onAction('monitor'); break;
      case 'KeyK': onAction('raw'); break;
      case 'KeyC': onAction('clip'); break;
      case 'KeyQ': onAction('quit'); break;
      case 'Slash': onAction('help'); break;
      // the chords a browser can never hand us — send them as commands instead
      case 'Delete': send({ t: 'key', k: 'delete', mods: ['ctrl', 'alt'] }); break;
      case 'Tab': send({ t: 'key', k: 'tab', mods: ['alt'] }); break;
      case 'KeyS': send({ t: 'key', k: 'cmd', mods: [] }); break;
      case 'Escape': send({ t: 'key', k: 'escape', mods: [] }); break;
      default: menuUsed = e.code !== 'ControlRight'; break;
    }
  }

  // ── keyboard lock (Chromium, fullscreen only) ─────────────────────────────
  // With it, Ctrl+W / Ctrl+T / Alt+Tab / Escape reach this page instead of the
  // browser or the OS. Without it those stay local — the chords above cover the
  // ones that matter.
  async function onFullscreenChange() {
    const fs = !!document.fullscreenElement;
    try {
      if (fs && navigator.keyboard && navigator.keyboard.lock) {
        await navigator.keyboard.lock();
        keyboardLocked = true;
      } else if (!fs && navigator.keyboard && navigator.keyboard.unlock) {
        navigator.keyboard.unlock();
        keyboardLocked = false;
      }
    } catch { keyboardLocked = false; }
    onAction('capture');
  }

  function clamp01(v) { return v < 0 ? 0 : v > 1 ? 1 : v; }
  function round6(v) { return Math.round(v * 1e6) / 1e6; }

  return {
    config, take, release, toggle,
    driving: () => driving,
    setKeyHold: (v) => { keyHold = !!v; },
    setRaw: (v) => { releaseAll(); raw = !!v; },
    raw: () => raw,
    keyboardLocked: () => keyboardLocked,
    // place the pointer where the user clicked before control was taken
    warpFromClient(clientX, clientY) {
      const r = view.rect();
      const box = stage.getBoundingClientRect();
      if (!r.w || !r.h) return;
      nx = clamp01((clientX - box.left - r.ox) / r.w);
      ny = clamp01((clientY - box.top - r.oy) / r.h);
      view.setCursor(nx, ny, true);
      send({ t: 'mvabs', nx: round6(nx), ny: round6(ny) });
    },
    center() { nx = .5; ny = .5; view.setCursor(nx, ny, driving); },
    releaseAll,
  };
})();

File: desktop/input.go
package main

import (
	"math"
	"time"
	"unicode"

	"github.com/hajimehoshi/ebiten/v2"
	"github.com/hajimehoshi/ebiten/v2/inpututil"
	"github.com/zlef-fr/zlefremote/desktop/internal/core"
)

// InputPump turns this machine's mouse and keyboard into host commands.
//
// Pointer: while grabbed the OS cursor is *captured* (ebiten reports unbounded
// virtual coordinates), so we keep our own cursor position, clamp it to the
// remote screen and send it as a normalized {"t":"mvabs"}. That gives absolute,
// self-correcting positioning — a dropped packet can't leave the two pointers
// permanently offset the way relative motion would — while the pointer can
// never escape the window mid-drag.
//
// Keyboard: characters and chords travel by different routes, see keymap.go.
// Modifiers are forwarded LAZILY: a modifier is only pressed on the host right
// before a chord or a click needs it, because the AltGr/Shift composition that
// produces a character already happened on this keyboard — holding those
// modifiers on the host too would corrupt the injected text.
type InputPump struct {
	send  func(any)
	prefs *core.Prefs

	grabbed  bool
	keyHold  bool // host supports kdown/kup (agent ≥ 1.7.0)
	rawKeys  bool
	screenW  int
	screenH  int
	viewRect rect

	// virtual pointer, normalized 0..1 over the viewed display
	nx, ny   float64
	lastCur  [2]int
	haveCur  bool
	lastSent time.Time

	sentMods    map[string]bool
	heldKeys    map[string]bool
	heldButtons map[string]bool

	scrollX, scrollY float64

	hotkeyUsed bool
}

type rect struct{ x, y, w, h float64 }

// hotkey actions the app must handle (everything the pump can't do itself).
const (
	actFullscreen  = "fullscreen"
	actPreset      = "core.Preset"
	actMonitor     = "monitor"
	actRawKeyboard = "raw"
	actDisconnect  = "disconnect"
	actHUD         = "hud"
	actClipPush    = "clip"
	actGrabOn      = "grab-on"
	actGrabOff     = "grab-off"
	actHelp        = "help"
)

func NewInputPump(send func(any), prefs *core.Prefs) *InputPump {
	return &InputPump{
		send:        send,
		prefs:       prefs,
		nx:          .5,
		ny:          .5,
		sentMods:    map[string]bool{},
		heldKeys:    map[string]bool{},
		heldButtons: map[string]bool{},
	}
}

func (p *InputPump) SetHostScreen(w, h int) { p.screenW, p.screenH = w, h }
func (p *InputPump) SetKeyHold(v bool)      { p.keyHold = v }
func (p *InputPump) SetView(r rect)         { p.viewRect = r }
func (p *InputPump) Grabbed() bool          { return p.grabbed }
func (p *InputPump) Pointer() (float64, float64) {
	return p.nx, p.ny
}

// Grab starts/stops forwarding. Releasing everything on the way out is not
// optional: a modifier or button left down on the host would keep "typing" long
// after the user looked away.
func (p *InputPump) Grab(on bool) {
	if p.grabbed == on {
		return
	}
	p.grabbed = on
	if on {
		ebiten.SetCursorMode(ebiten.CursorModeCaptured)
		p.haveCur = false
	} else {
		ebiten.SetCursorMode(ebiten.CursorModeVisible)
		p.ReleaseAll()
	}
}

// ReleaseAll lifts every key/button we are holding on the host.
func (p *InputPump) ReleaseAll() {
	for b := range p.heldButtons {
		p.send(map[string]any{"t": "up", "b": b})
		delete(p.heldButtons, b)
	}
	for k := range p.heldKeys {
		if p.keyHold {
			p.send(map[string]any{"t": "kup", "k": k})
		}
		delete(p.heldKeys, k)
	}
	p.releaseMods()
}

// Update runs once per tick and returns the local hotkey actions it recognised.
func (p *InputPump) Update() []string {
	var actions []string

	// The window losing focus (alt-tab away, screen lock) must never leave keys
	// stuck on the other machine.
	if !ebiten.IsFocused() {
		if p.grabbed {
			p.Grab(false)
			actions = append(actions, actGrabOff)
		}
		return actions
	}

	if acts, handled := p.hotkeys(); handled {
		return append(actions, acts...)
	}
	if !p.grabbed {
		return actions
	}

	p.pointer()
	p.buttons()
	p.wheel()
	p.keyboard()
	return actions
}

// hotkeys implements the Right-Ctrl "menu key" chords. While it is held nothing
// reaches the host; a tap on its own toggles the grab.
func (p *InputPump) hotkeys() ([]string, bool) {
	held := ebiten.IsKeyPressed(hotkeyModifier)
	if inpututil.IsKeyJustPressed(hotkeyModifier) {
		p.hotkeyUsed = false
		// drop anything currently held so the chord can't stick
		p.ReleaseAll()
	}
	if !held {
		if inpututil.IsKeyJustReleased(hotkeyModifier) && !p.hotkeyUsed {
			p.Grab(!p.grabbed)
			if p.grabbed {
				return []string{actGrabOn}, true
			}
			return []string{actGrabOff}, true
		}
		return nil, false
	}

	var out []string
	for _, k := range inpututil.AppendJustPressedKeys(nil) {
		// A modifier is never a chord by itself. This also covers the extra
		// keys ebiten reports alongside the physical one (the "either side"
		// aliases, and on X11 a Right-Ctrl press that also surfaces as
		// ControlLeft) — counting any of them would stop a bare tap of the
		// menu key from toggling the grab.
		if _, isMod := modifierKeys[k]; isMod || k == ebiten.KeyControl ||
			k == ebiten.KeyShift || k == ebiten.KeyAlt || k == ebiten.KeyMeta {
			continue
		}
		switch k {
		case ebiten.KeyF:
			out = append(out, actFullscreen)
		case ebiten.KeyQ:
			out = append(out, actDisconnect)
		case ebiten.KeyM:
			out = append(out, actMonitor)
		case ebiten.KeyP:
			out = append(out, actPreset)
		case ebiten.KeyK:
			out = append(out, actRawKeyboard)
		case ebiten.KeyH:
			out = append(out, actHUD)
		case ebiten.KeyC:
			out = append(out, actClipPush)
		case ebiten.KeySlash:
			out = append(out, actHelp)
		case ebiten.KeyDelete:
			// the one chord no OS lets an app receive: send it to the host
			p.sendChord("delete", []string{"ctrl", "alt"})
		case ebiten.KeyTab:
			p.sendChord("tab", []string{"alt"})
		case ebiten.KeyS:
			p.sendChord("cmd", nil)
		case ebiten.KeyEscape:
			p.sendChord("escape", nil)
		}
		p.hotkeyUsed = true
	}
	return out, true
}

// sendChord presses a key with modifiers as one atomic action on the host.
func (p *InputPump) sendChord(key string, mods []string) {
	p.send(map[string]any{"t": "key", "k": key, "mods": mods})
}

func (p *InputPump) pointer() {
	cx, cy := ebiten.CursorPosition()
	if !p.haveCur {
		p.lastCur = [2]int{cx, cy}
		p.haveCur = true
		return
	}
	dx, dy := float64(cx-p.lastCur[0]), float64(cy-p.lastCur[1])
	p.lastCur = [2]int{cx, cy}
	if dx == 0 && dy == 0 {
		return
	}
	sens := p.prefs.Sensitivity
	if sens <= 0 {
		sens = 1
	}
	// One pixel of local motion moves the remote pointer by one *remote* pixel:
	// the view is a scaled copy of the host screen, so divide by the view size.
	if p.viewRect.w > 0 && p.viewRect.h > 0 {
		p.nx += dx * sens / p.viewRect.w
		p.ny += dy * sens / p.viewRect.h
	}
	p.nx = clamp01(p.nx)
	p.ny = clamp01(p.ny)

	// The host injects at most one move per frame anyway; 120 Hz is plenty and
	// keeps a fast flick from flooding a mobile uplink.
	if time.Since(p.lastSent) < 8*time.Millisecond {
		return
	}
	p.lastSent = time.Now()
	p.send(map[string]any{"t": "mvabs", "nx": round6(p.nx), "ny": round6(p.ny)})
}

func (p *InputPump) buttons() {
	for btn, name := range map[ebiten.MouseButton]string{
		ebiten.MouseButtonLeft:   "left",
		ebiten.MouseButtonRight:  "right",
		ebiten.MouseButtonMiddle: "middle",
	} {
		if inpututil.IsMouseButtonJustPressed(btn) {
			p.syncMods()
			// pin the pointer before the press: a click must land where the
			// user sees the cursor even if the last move was throttled away.
			p.send(map[string]any{"t": "mvabs", "nx": round6(p.nx), "ny": round6(p.ny)})
			p.send(map[string]any{"t": "down", "b": name})
			p.heldButtons[name] = true
		}
		if inpututil.IsMouseButtonJustReleased(btn) {
			p.send(map[string]any{"t": "up", "b": name})
			delete(p.heldButtons, name)
		}
	}
}

func (p *InputPump) wheel() {
	dx, dy := ebiten.Wheel()
	if dx == 0 && dy == 0 {
		return
	}
	speed := p.prefs.ScrollSpeed
	if speed <= 0 {
		speed = 1
	}
	if p.prefs.Natural {
		dy = -dy
	}
	// ebiten reports notches (±1 per detent, fractional on touchpads); robotgo
	// scrolls in "clicks", so 3 lines per notch matches a normal desktop.
	p.scrollX += dx * speed * 3
	p.scrollY += dy * speed * 3
	ix, iy := int(p.scrollX), int(p.scrollY)
	if ix == 0 && iy == 0 {
		return
	}
	p.scrollX -= float64(ix)
	p.scrollY -= float64(iy)
	p.send(map[string]any{"t": "scroll", "dx": ix, "dy": iy})
}

func (p *InputPump) keyboard() {
	pressed := inpututil.AppendJustPressedKeys(nil)
	released := inpututil.AppendJustReleasedKeys(nil)
	chars := ebiten.AppendInputChars(nil)

	if p.rawKeys {
		p.rawKeyboard(pressed, released)
		return
	}

	// composed characters first: they are what the user actually typed
	if s := printable(chars); s != "" {
		p.releaseMods() // never type through a held modifier (see type above)
		p.send(map[string]any{"t": "text", "s": s})
	}

	for _, k := range pressed {
		if _, isMod := modifierKeys[k]; isMod {
			continue // lazy: pressed on demand by syncMods
		}
		if textKeys[k] && !p.modsActive() {
			continue // plain character: already handled by the text path
		}
		name, ok := keyName(k)
		if !ok {
			continue
		}
		p.syncMods()
		if p.keyHold {
			p.send(map[string]any{"t": "kdown", "k": name})
			p.heldKeys[name] = true
		} else {
			// older agent: no hold/release, tap with the live modifiers
			p.send(map[string]any{"t": "key", "k": name, "mods": p.activeMods()})
		}
	}

	for _, k := range released {
		if _, isMod := modifierKeys[k]; isMod {
			p.syncModsRelease(k)
			continue
		}
		name, ok := keyName(k)
		if !ok || !p.heldKeys[name] {
			continue
		}
		delete(p.heldKeys, name)
		if p.keyHold {
			p.send(map[string]any{"t": "kup", "k": name})
		}
	}
}

// rawKeyboard forwards physical keys verbatim, modifiers included, and never
// sends composed text. Right for games and for remote-controlling a machine
// whose layout matches this one; wrong for typing across layouts.
func (p *InputPump) rawKeyboard(pressed, released []ebiten.Key) {
	for _, k := range pressed {
		if name, ok := keyName(k); ok {
			if p.keyHold {
				p.send(map[string]any{"t": "kdown", "k": name})
				p.heldKeys[name] = true
			} else {
				p.send(map[string]any{"t": "key", "k": name, "mods": p.activeMods()})
			}
		}
	}
	for _, k := range released {
		if name, ok := keyName(k); ok {
			delete(p.heldKeys, name)
			if p.keyHold {
				p.send(map[string]any{"t": "kup", "k": name})
			}
		}
	}
}

// modsActive reports whether a chord-forming modifier (not Shift, which only
// changes the character) is held locally.
func (p *InputPump) modsActive() bool {
	return ebiten.IsKeyPressed(ebiten.KeyControlLeft) ||
		ebiten.IsKeyPressed(ebiten.KeyAltLeft) || ebiten.IsKeyPressed(ebiten.KeyAltRight) ||
		ebiten.IsKeyPressed(ebiten.KeyMetaLeft) || ebiten.IsKeyPressed(ebiten.KeyMetaRight)
}

// activeMods is the live modifier list for the tap fallback.
func (p *InputPump) activeMods() []string {
	var m []string
	if ebiten.IsKeyPressed(ebiten.KeyShiftLeft) || ebiten.IsKeyPressed(ebiten.KeyShiftRight) {
		m = append(m, "shift")
	}
	if ebiten.IsKeyPressed(ebiten.KeyControlLeft) {
		m = append(m, "ctrl")
	}
	if ebiten.IsKeyPressed(ebiten.KeyAltLeft) || ebiten.IsKeyPressed(ebiten.KeyAltRight) {
		m = append(m, "alt")
	}
	if ebiten.IsKeyPressed(ebiten.KeyMetaLeft) || ebiten.IsKeyPressed(ebiten.KeyMetaRight) {
		m = append(m, "meta")
	}
	return m
}

// syncMods presses on the host exactly the modifiers held here, right before an
// event that needs them.
func (p *InputPump) syncMods() {
	if !p.keyHold {
		return
	}
	want := map[string]bool{}
	for _, m := range p.activeMods() {
		if m == "meta" {
			m = "cmd"
		}
		want[m] = true
	}
	for m := range want {
		if !p.sentMods[m] {
			p.send(map[string]any{"t": "kdown", "k": m})
			p.sentMods[m] = true
		}
	}
	for m := range p.sentMods {
		if !want[m] {
			p.send(map[string]any{"t": "kup", "k": m})
			delete(p.sentMods, m)
		}
	}
}

// syncModsRelease lifts a modifier on the host when it is released here (only
// if we had actually pressed it).
func (p *InputPump) syncModsRelease(k ebiten.Key) {
	name := modifierKeys[k]
	if name == "" || !p.sentMods[name] {
		return
	}
	// left/right variants share one host name — wait until both are up
	switch name {
	case "shift":
		if ebiten.IsKeyPressed(ebiten.KeyShiftLeft) || ebiten.IsKeyPressed(ebiten.KeyShiftRight) {
			return
		}
	case "ctrl":
		if ebiten.IsKeyPressed(ebiten.KeyControlLeft) {
			return
		}
	case "alt":
		if ebiten.IsKeyPressed(ebiten.KeyAltLeft) || ebiten.IsKeyPressed(ebiten.KeyAltRight) {
			return
		}
	case "cmd":
		if ebiten.IsKeyPressed(ebiten.KeyMetaLeft) || ebiten.IsKeyPressed(ebiten.KeyMetaRight) {
			return
		}
	}
	p.send(map[string]any{"t": "kup", "k": name})
	delete(p.sentMods, name)
}

func (p *InputPump) releaseMods() {
	for m := range p.sentMods {
		if p.keyHold {
			p.send(map[string]any{"t": "kup", "k": m})
		}
		delete(p.sentMods, m)
	}
}

// SetRawKeyboard switches typing modes, releasing whatever the old mode held.
func (p *InputPump) SetRawKeyboard(v bool) {
	if p.rawKeys == v {
		return
	}
	p.ReleaseAll()
	p.rawKeys = v
}

func (p *InputPump) RawKeyboard() bool { return p.rawKeys }

// SetPointer re-centres the virtual pointer (used when the viewed monitor
// changes, so the cursor doesn't jump to a stale corner).
func (p *InputPump) SetPointer(nx, ny float64) {
	p.nx, p.ny = clamp01(nx), clamp01(ny)
}

// WarpTo moves the host pointer to a normalized point right away. Taking
// control by clicking in the picture uses it, so the remote cursor lands under
// the local one instead of wherever it was left.
func (p *InputPump) WarpTo(nx, ny float64) {
	p.SetPointer(nx, ny)
	p.send(map[string]any{"t": "mvabs", "nx": round6(p.nx), "ny": round6(p.ny)})
}

func printable(rs []rune) string {
	out := make([]rune, 0, len(rs))
	for _, r := range rs {
		if r == '\n' || r == '\r' || r == '\t' || r == 0x7f {
			continue // these travel as key events, not as text
		}
		if unicode.IsControl(r) {
			continue
		}
		out = append(out, r)
	}
	return string(out)
}

func clamp01(v float64) float64 {
	if v < 0 {
		return 0
	}
	if v > 1 {
		return 1
	}
	return v
}

// round6 keeps the normalized pointer accurate to the pixel on a 4K screen
// (1/4096 needs ~5 decimals) while keeping frames small.
func round6(v float64) float64 { return math.Round(v*1e6) / 1e6 }
